crates/bevy_pbr/src/meshlet/asset.rs

```diff
@@ -68,84 +70,141 @@ pub struct MeshletBoundingSpheres {
+        writer: &'a mut Writer,
+        asset: SavedAsset<'a, MeshletMesh>,
+        _settings: &'a (),
+    ) -> Result<(), MeshletMeshSaveOrLoadError> {
+        // Compress asset data
+        let mut compressed_asset_data = Vec::new();
+        let mut compressor = FrameEncoder::new(&mut compressed_asset_data);
+        write_slice(&asset.vertex_data, &mut compressor)?;
+        write_slice(&asset.vertex_ids, &mut compressor)?;
+        write_slice(&asset.indices, &mut compressor)?;
+        write_slice(&asset.meshlets, &mut compressor)?;
+        write_slice(&asset.bounding_spheres, &mut compressor)?;
+        compressor.finish()?;
+
+        // Write asset metadata
+        writer
+            .write_all(&MESHLET_MESH_ASSET_VERSION.to_le_bytes())
+            .await?;
+        writer
+            .write_all(&asset.worst_case_meshlet_triangles.to_le_bytes())
+            .await?;
+        writer
+            .write_all(&(compressed_asset_data.len() as u64).to_le_bytes())
+            .await?;
 
-        Ok(asset)
-    }
+        // Write compressed asset data
+        writer.write_all(&compressed_asset_data).await?;
 
-    fn extensions(&self) -> &[&str] {
-        &["meshlet_mesh"]
+        Ok(())
     }
 }
 
-impl AssetSaver for MeshletMeshSaverLoad {
+impl AssetLoader for MeshletMeshSaverLoader {
     type Asset = MeshletMesh;
     type Settings = ();
-    type OutputLoader = Self;
     type Error = MeshletMeshSaveOrLoadError;
 
-    async fn save<'a>(
+    async fn load<'a>(
         &'a self,
-        writer: &'a mut Writer,
-        asset: SavedAsset<'a, Self::Asset>,
-        _settings: &'a Self::Settings,
-    ) -> Result<(), Self::Error> {
-        writer
-            .write_all(&MESHLET_MESH_ASSET_VERSION.to_le_bytes())
-            .await?;
+        reader: &'a mut dyn Reader,
+        _settings: &'a (),
+        _load_context: &'a mut LoadContext<'_>,
+    ) -> Result<MeshletMesh, MeshletMeshSaveOrLoadError> {
+        // Load and check asset version
+        let version = async_read_u64(reader).await?;
+        if version != MESHLET_MESH_ASSET_VERSION {
+            return Err(MeshletMeshSaveOrLoadError::WrongVersion { found: version });
+        }
 
-        let mut bytes = Vec::new();
-        let mut sync_writer = FrameEncoder::new(&mut bytes);
-        bincode::serialize_into(&mut sync_writer, asset.get())?;
-        sync_writer.finish()?;
-        writer.write_all(&bytes).await?;
+        // Load asset metadata
+        let worst_case_meshlet_triangles = async_read_u64(reader).await?;
+        let compressed_asset_data_len = async_read_u64(reader).await? as usize;
+
+        // Load compressed asset data
+        let mut compressed_asset_data = Vec::with_capacity(compressed_asset_data_len);
+        reader.read_to_end(&mut compressed_asset_data).await?;
+
+        // Convert compressed data back to an asset
+        let reader = &mut FrameDecoder::new(Cursor::new(compressed_asset_data));
```

Comment:
Does LZ4 compression actually help? You might have to use delta encoding (i.e. instead of storing indices you'd store the wrapping difference from the previous index) if you want it to have much impact

Reply:
It did when I used bincode, I should probably double check again. 

Nanite has a whole bunch of tricks they use for compressing disk and in-memory asset data, but they're low priority for me vs runtime performance at the moment, and I don't really have much experience with this field https://advances.realtimerendering.com/s2021/Karis_Nanite_SIGGRAPH_Advances_2021_final.pdf.

Reply:
9.73mb without compression, 4.97mb with.